### PR TITLE
[13.0][MIG] sale_order_type

### PR DESCRIPTION
**Pull request by kongrattapong**
Ready to review

### CODE REVIEW

**Review by marylla — APPROVED** [review]
Hi there, 
I made a functional review. I could create a new sale order type with a different sequence, payment term, shipping policy, and so on.
Quotations were created according to the chosen type. So, LGTM. :)

File: sale_order_type/models/sale.py [hunk 1]
```diff
@@ -23,20 +22,23 @@ def onchange_partner_id(self):
...
+                    or order.warehouse_id.id,
+                    "picking_policy": order.type_id.picking_policy
+                    or order.picking_policy,
+                    "payment_term_id": order.type_id.payment_term_id.id
+                    or order.payment_term_id,
+                    "pricelist_id": order.type_id.pricelist_id.id or order.pricelist_id,
+                    "incoterm": order.type_id.incoterm_id.id or order.incoterm,
+                }
+            )
+            order.write(vals)
```
Comment (rousseldenis) [inline sale_order_type/models/sale.py]: The good method is not doing a write() in an onchange.

Either you affect each value (like in the code you deleted), but it has disadvantage to do a write after each affectation (and an SQL query).

Either (the best) you use update() on the record:

```python
order.update({
                    "warehouse_id": order.type_id.warehouse_id.id
                    or order.warehouse_id.id,
                    "picking_policy": order.type_id.picking_policy
                    or order.picking_policy,
                    "payment_term_id": order.type_id.payment_term_id.id
                    or order.payment_term_id,
                    "pricelist_id": order.type_id.pricelist_id.id or order.pricelist_id,
                    "incoterm": order.type_id.incoterm_id.id or order.incoterm,
                })
```

Moreover the way you did the update is a llitle bit tricky. Don't reaffect the order field value if the type value does not contain one, but build your dict with value or not.

So, I would have done:

```python
for order in self:
    vals = {}
    order_type = order.type_id
    if order_type.warehouse_id:
        vals.update({"warehouse_id": o_type.warehouse_id})
    if order_type.picking_policy:
        vals.update({"picking_policy": o_type.picking_policy})
    if order_type.payment_term_id:
        vals.update({"payment_term_id": o_type.payment_term_id})
    if order_type.pricelist_id:
        vals.update({"pricelist_id": o_type.pricelist_id})
    if order_type.incoterm_id:
        vals.update({"incoterm": order_type.incoterm_id})
    
    if vals:
        order.update(vals)
```

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

File: sale_order_type/models/sale.py (hunk 1 above)
Comment (kongrattapong) [reply sale_order_type/models/sale.py]: @rousseldenis I fixed as your advice, thank you for review.

**Review by rousseldenis — APPROVED** [review]
Code review

**Comment by sergio-teruel** [comment]
This module contains more than one module (sale_order_line_description)
@kongrattapong Can you do the PR only with sale_order_type module?
ping @rousseldenis

**Comment by pedrobaeza** [comment]
Aborting because there are changes of other modules.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Remove the changes in sale_order_line_description module.

**Comment by pedrobaeza** [comment]
@dreispt please always check the full thread, not only the approvals. Some people thinks that putting request changes is very hard and only comment in the thread, but this is not correct to be merged as is.

**Comment by kongrattapong** [comment]
> This module contains more than one module (sale_order_line_description)
> @kongrattapong Can you do the PR only with sale_order_type module?
> ping @rousseldenis

Thank for review I Removed.

File: sale_order_type/README.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    :target: https://odoo-community.org/page/development-status
+    :alt: Beta
+.. |badge2| image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+.. |badge3| image:: https://img.shields.io/badge/github-OCA%2Fsale--workflow-lightgray.png?logo=github
+    :target: https://github.com/OCA/sale-workflow/tree/13.0-mig-sale_order_type/sale_order_type
+    :alt: OCA/sale-workflow
+.. |badge4| image:: https://img.shields.io/badge/weblate-Translate%20me-F47D42.png
+    :target: https://translation.odoo-community.org/projects/sale-workflow-13-0-mig-sale_order_type/sale-workflow-13-0-mig-sale_order_type-sale_order_type
```
Comment (sergio-teruel) [inline sale_order_type/README.rst]: CAn you check this path?? I think that is not correctly

File: sale_order_type/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    @api.onchange("partner_id", "company_id")
+    def _onchange_partner_id(self):
+        res = super(AccountMove, self)._onchange_partner_id()
+        sale_type = (
+            self.partner_id.sale_type or self.partner_id.commercial_partner_id.sale_type
+        )
+        if sale_type:
+            self.sale_type_id = sale_type
+        return res
```
Comment (sergio-teruel) [inline sale_order_type/models/account_move.py]: Could you change onchange methods to computes with a depends??

File: sale_order_type/models/sale.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    _inherit = "sale.order"
+
+    def _get_order_type(self):
+        return self.env["sale.order.type"].search([], limit=1)
+
+    type_id = fields.Many2one(
+        comodel_name="sale.order.type", string="Type", default=_get_order_type
+    )
+
+    @api.onchange("partner_id")
```
Comment (sergio-teruel) [inline sale_order_type/models/sale.py]: Try change onchange methods to compute one..

File: sale_order_type/tests/test_sale_order_type.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            "product_uom_qty": 1.0,
+            "price_unit": self.product.lst_price,
+        }
+        return {"partner_id": self.partner.id, "order_line": [(0, 0, sale_line_dict)]}
+
+    def test_sale_order_confirm(self):
+        sale_type = self.sale_type
+        order_vals = self.get_sale_order_vals()
+        order = self.sale_order_model.create(order_vals)
+        order.onchange_partner_id()
```
Comment (sergio-teruel) [inline sale_order_type/tests/test_sale_order_type.py]: If you change onchange method to compute methods you can remove this calls

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by dreispt** [comment]
@pedrobaeza were your concerns fixed? Thanks!

**Comment by pedrobaeza** [comment]
My initial concerns are addressed, but I'm going to convert onchanges to compute methods for improving a lot the number of lines and the side effects. I'll do it today or tomorrow.

**Comment by pedrobaeza** [comment]
It turns out that there were more incorrect things in the migration, like outdated record rules, or incorrect payment term field name. Now everything should be OK.

/ocabot merge

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@sbidoul any idea why this is constant failing with no apparent "user error"?

**Comment by pedrobaeza** [comment]
Let's merge it manually as it's not possible to use bot.

**Pull request merged**